allennlp/common/file_utils.py

```diff
@@ -63,6 +65,38 @@
     )
 
 
+class FileLock(_FileLock):
+    """
+    This is just a subclass of the `FileLock` class from the `filelock` library, except that
+    it adds an additional argument to the `__init__` method: `read_only_ok`.
+
+    By default this flag is `False`, which an exception will be thrown when a lock
+    can't be acquired due to lack of write permissions.
+    But if this flag is set to `True`, a warning will be emitted instead of an error when
+    the lock already exists but the lock can't be acquired because write access is blocked.
+    """
+
+    def __init__(
+        self, lock_file: Union[str, PathLike], timeout=-1, read_only_ok: bool = False
+    ) -> None:
+        super().__init__(str(lock_file), timeout=timeout)
+        self._read_only_ok = read_only_ok
+
+    @overrides
+    def acquire(self, timeout=None, poll_interval=0.05):
+        try:
+            super().acquire(timeout=timeout, poll_intervall=poll_interval)
+        except PermissionError:
```

Comment:
@epwalsh  I don't think it catches all types of errors. For instance when running from a Docker I get
`OSError: [Errno 30] Read-only file system: '/opt/ml/model/vocabulary/.lock'`

Here is the traceback

```
Traceback (most recent call last):
  File "/opt/program/predictor.py", line 43, in get_model
    cls.model.initialize(model_path)
  File "/opt/program/SiameseEncoder.py", line 45, in initialize
    "model.transformer_model": siamese_base_path,
  File "/usr/local/lib/python3.6/site-packages/allennlp/models/archival.py", line 208, in load_archive
    model = _load_model(config.duplicate(), weights_path, serialization_dir, cuda_device)
  File "/usr/local/lib/python3.6/site-packages/allennlp/models/archival.py", line 246, in _load_model
    cuda_device=cuda_device,
  File "/usr/local/lib/python3.6/site-packages/allennlp/models/model.py", line 406, in load
    return model_class._load(config, serialization_dir, weights_file, cuda_device)
  File "/usr/local/lib/python3.6/site-packages/allennlp/models/model.py", line 294, in _load
    vocab_dir, vocab_params.get("padding_token"), vocab_params.get("oov_token")
  File "/usr/local/lib/python3.6/site-packages/allennlp/data/vocabulary.py", line 339, in from_files
    with FileLock(os.path.join(directory, ".lock"), read_only_ok=True):
  File "/usr/local/lib/python3.6/site-packages/filelock.py", line 323, in __enter__
    self.acquire()
  File "/usr/local/lib/python3.6/site-packages/allennlp/common/file_utils.py", line 88, in acquire
    super().acquire(timeout=timeout, poll_intervall=poll_interval)
  File "/usr/local/lib/python3.6/site-packages/filelock.py", line 271, in acquire
    self._acquire()
  File "/usr/local/lib/python3.6/site-packages/filelock.py", line 384, in _acquire
    fd = os.open(self._lock_file, open_mode)
```

Reply:
Ah, good catch, thank you! Just updated to catch `OSError` as well.